pkg/controller/certificatemanager/certificatemanager.go

```diff
@@ -237,44 +240,84 @@ func (cm *certificateManager) getKeyPair(cli client.Client, secretName, secretNa
 		return nil, nil, err
 	}
 
+	// First, check that it's not expired.
 	if x509Cert.NotAfter.Before(time.Now()) || x509Cert.NotBefore.After(time.Now()) {
 		if !readCertOnly && strings.HasPrefix(x509Cert.Issuer.CommonName, rmeta.TigeraOperatorCAIssuerPrefix) {
 			if cm.keyPair.CertificateManagement != nil {
 				// When certificate management is enabled, we can simply return a certificate management key pair;
 				// the old secret will be deleted automatically.
 				return certificateManagementKeyPair(cm, secretName, nil), nil, nil
 			}
+
 			// We return nil, so a new secret will be created for expired (legacy) operator signed secrets.
 			return nil, nil, nil
 		}
-		// We return an error for byo secrets.
+
+		// The certificate was not issued by tigera-operator, or the caller indicated that certs shouldn't be modified.
+		// Consider this a BYO certificate, and error out rather than generate a new certificate.
 		return nil, nil, fmt.Errorf("secret %s is not valid at this date", secretName)
 	}
 	var issuer certificatemanagement.KeyPairInterface
 	if x509Cert.Issuer.CommonName == rmeta.TigeraOperatorCAIssuerPrefix {
 		if cm.keyPair.CertificateManagement != nil {
 			return certificateManagementKeyPair(cm, secretName, nil), nil, nil
 		}
+
+		// This certificate wasn't issued using external certificate management.
+		if !ValidForClientAndServer(x509Cert) {
+			// The certificate isn't valid as both a client and server cert. Older versions of the operator
+			// generated certificates that were only valid for use as a server certificate. However, some components now
+			// use these as client certificates in mTLS connections as well. So, regenerate any that need this capability.
+			return nil, nil, nil
```

Comment:
Should we be returning an error here to convey why the certificate isn't valid?

Reply:
That's right.